internal/credentials/ecr/ecr.go

```diff
@@ -0,0 +1,151 @@
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ecr"
+	"github.com/patrickmn/go-cache"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	regionKey = "awsRegion"
+	idKey     = "awsAccessKeyID"
+	secretKey = "awsSecretAccessKey"
+)
+
+// CredentialHelper is an interface for components that can extract a username
+// and password from a Secret containing an AWS region, access key id, and
+// secret access key.
+type CredentialHelper interface {
+	// GetUsernameAndPassword extracts username and password (a token that lives
+	// for 12 hours) from a Secret IF the Secret contains an AWS region, access
+	// key id, and secret access key. If the Secret does not contain ANY of these
+	// fields, this function will return empty strings and a nil error. If the
+	// Secret contains some but not all of these fields, this function will return
+	// an error. Implementations may cache the token for efficiency.
+	GetUsernameAndPassword(*corev1.Secret) (string, string, error)
+}
+
+type credentialHelper struct {
+	tokenCache *cache.Cache
+
+	// The following behaviors are overridable for testing purposes:
+
+	getAuthTokenFn func(string, string, string) (string, error)
+}
+
+// NewCredentialHelper returns an implementation of the CredentialHelper
+// interface that utilizes a cache to avoid unnecessary calls to AWS.
+func NewCredentialHelper() CredentialHelper {
+	return &credentialHelper{
+		tokenCache: cache.New(
+			// Tokens live for 12 hours. We'll hang on to them for 10.
+			10*time.Hour, // Default ttl for each entry
+			time.Hour,    // Cleanup interval
+		),
+		getAuthTokenFn: getAuthToken,
+	}
+}
+
+// GetUsernameAndPassword implements the CredentialHelper interface.
+func (c *credentialHelper) GetUsernameAndPassword(
+	secret *corev1.Secret,
+) (string, string, error) {
+	region := string(secret.Data[regionKey])
+	accessKeyID := string(secret.Data[idKey])
+	secretAccessKey := string(secret.Data[secretKey])
+	var awsCount int
+	if region != "" {
+		awsCount++
+	}
+	if accessKeyID != "" {
+		awsCount++
+	}
+	if secretAccessKey != "" {
+		awsCount++
+	}
+	if awsCount == 0 {
+		return "", "", nil
+	}
+	if awsCount > 0 && awsCount < 3 {
+		return "", "", fmt.Errorf(
+			"%s, %s, and %s must all be set or all be unset",
+			regionKey, idKey, secretKey,
+		)
+	}
+	return c.getUsernameAndPassword(region, accessKeyID, secretAccessKey)
+}
+
+func (c *credentialHelper) getUsernameAndPassword(
+	region, accessKeyID, secretAccessKey string,
+) (string, string, error) {
+	cacheKey := tokenCacheKey(region, accessKeyID, secretAccessKey)
+
+	if entry, exists := c.tokenCache.Get(cacheKey); exists {
+		return decodeAuthToken(entry.(string)) // nolint: forcetypeassert
+	}
+
+	encodedToken, err := c.getAuthTokenFn(region, accessKeyID, secretAccessKey)
+	if err != nil {
+		return "", "", fmt.Errorf("error getting ECR auth token: %w", err)
+	}
+
+	// Cache the encoded token
+	c.tokenCache.Set(cacheKey, encodedToken, cache.DefaultExpiration)
```

Comment:
`DefaultExperiation` equals to forever. Shouldn't this be a little bit less than the TTL of the token itself (e.g. 10h)?

Reply:
I don't `cache.DefaultExpiration` believe it does equal forever.

https://github.com/patrickmn/go-cache/blob/46f407853014144407b6c2ec7ccc76bf67958d93/cache.go#L29-L32

```golang
// For use with functions that take an expiration time. Equivalent to
// passing in the same expiration duration as was given to New() or
// NewFrom() when the cache was created (e.g. 5 minutes.)
DefaultExpiration time.Duration = 0
```

The implementation of the `Set()` function also supports the accuracy of that comment.

https://github.com/patrickmn/go-cache/blob/46f407853014144407b6c2ec7ccc76bf67958d93/cache.go#L54-L56

```golang
if d == DefaultExpiration {
	d = c.defaultExpiration
}
```

And I did indeed set the default expiration for `c.tokenCache` to `10*time.Hour`:

```golang
tokenCache: cache.New(
	// Tokens live for 12 hours. We'll hang on to them for 10.
	10*time.Hour, // Default ttl for each entry
	time.Hour,    // Cleanup interval
)
```